content/posts/2022-09-14-introducing-jaiminho/index.md

```diff
@@ -0,0 +1,136 @@
+
+
+# The Transactional outbox pattern
+
+For simplicity purposes, let’s assume the common case of an application that has to save data to a database and notify it to an event stream. The idea behind the transactional outbox pattern is combining both writes into a single atomic transaction, which guarantees that if one of them fails, the entire transaction fails, avoiding inconsistent scenarios; in this case, it means that if a record is persisted into the database, it is guaranteed to be notified to the stream. The pattern achieves this goal by introducing two concepts:
+
+- Outbox table
+
+When storing a record to the database, the corresponding event should be persisted into an outbox table, and both inserts should be in the same transaction. With this strategy we can leverage the atomicity property of relational databases transactions to ensure either both or no inserts will be successful.
+
+- Message relay
+
+The outbox table is guaranteed to have all events that should be notified, so we just have to ensure that they are really sent. This responsibility goes to the message relay, which will publish all events in the table to the stream and mark them as sent.
+Note: Since the message relay has to send an event and then mark it as sent, that characterizes another dual write, because if the relay fails to update the status after sending the event, it will be sent again on the next execution. So, to ensure no duplicated events are delivered, the message consumers must implement idempotence on their side.
+
+# Using Jaiminho in your project
+
+To add Jaiminho to your project, you have to:
+
+1) Install it:
+
+```console
+python -m pip install jaiminho
+```
+
+2) Add it to the Django project’s `INSTALLED_APPS`:
+
+```python
+INSTALLED_APPS = [
+    "django.contrib.auth",
+    "django.contrib.contenttypes",
+    ...
+    ...
+    ...
+    "jaiminho"
+]
+```
+
+3) Configure the options in the project’s Django `settings.py`:
+
+```python
+JAIMINHO_CONFIG = {
+    "PERSIST_ALL_EVENTS": False,
+    "DELETE_AFTER_SEND": True,
+    "PUBLISH_STRATEGY: "publish-on-commit"
+    }
+```
+
+4) Add the `@save_to_outbox` decorator to any method responsible for communicating with external systems (brokers, external APIs, etc):
+
+```python
+from jaiminho.send import save_to_outbox
+
+@save_to_outbox
+def any_external_call(**kwargs):
+    # do something
+    return
+```
+
+
+# How Jaiminho works
+
+With the default configuration, when you add the `@save_to_outbox` decorator to a method, whenever it is called and raises an exception, Jaiminho will create a database entry storing:
```

Comment:
If we use the keep order option, we save it always and store it for sending it in order. Right @antunesleo  ?

Reply:
Yeah! With the keep order option, the persist all events is mandatory. It is explained on the Persist all events subsection on the additional settings section.